File: sdks/python/nroutersdk/memory.py
"""Client-side conversation memory.

The gateway is stateless between calls. This module stores messages locally and
returns copies that callers can pass as ``messages``.
"""

from __future__ import annotations

import asyncio
import copy
from typing import Any, Protocol

from nroutersdk._errors import nRouterRequestError

ChatMessage = dict[str, Any]
_ROLES = {"system", "user", "assistant", "tool", "developer"}
_TENANCY_KEYS = {"organizationid", "orgid", "teamid", "userid", "nrouterorg"}


class MemoryStore(Protocol):
    def load(self) -> list[ChatMessage]: ...
    def save(self, messages: list[ChatMessage]) -> None: ...


class ArrayMemoryStore:
    """In-process memory store. It writes nowhere by default."""

    def __init__(self, seed: list[ChatMessage] | None = None) -> None:
        self._messages = [_clone_message(m) for m in (seed or [])]

    def load(self) -> list[ChatMessage]:
        return [_clone_message(m) for m in self._messages]

    def save(self, messages: list[ChatMessage]) -> None:
        self._messages = [_clone_message(m) for m in messages]


def sliding_window(
    messages: list[ChatMessage],
    max_messages: int | None = None,
    preserve_system: bool = True,
) -> list[ChatMessage]:
    """Prune a message list to the most recent `max_messages`, preserving the index 0
    system/developer message by default.
    """
    if max_messages is None:
        return [_clone_message(m) for m in messages]
    if max_messages <= 0:
        return []
    if len(messages) <= max_messages:
        return [_clone_message(m) for m in messages]
    if preserve_system and messages and messages[0].get("role") in ("system", "developer"):
        if max_messages == 1:
            return [_clone_message(messages[-1])]
        tail_count = max_messages - 1
        tail = [_clone_message(m) for m in messages[-tail_count:]]
        return [_clone_message(messages[0])] + tail
    return [_clone_message(m) for m in messages[-max_messages:]]


class Memory:
    """A small async-safe wrapper around a message store."""

    def __init__(
        self,
        store: MemoryStore | None = None,
        max_messages: int | None = None,
        preserve_system: bool = True,
    ) -> None:
        self._store = store or ArrayMemoryStore()
        self._lock = asyncio.Lock()
        self._max_messages = max_messages
        self._preserve_system = preserve_system

    async def add(self, message: ChatMessage) -> None:
        clean = _validate_message(message, "add()")
        async with self._lock:
            current = self._read()
            current.append(clean)
            self._write(current)

    async def messages(
        self,
        max_messages: int | None = None,
        preserve_system: bool | None = None,
    ) -> list[ChatMessage]:
        async with self._lock:
            msgs = self._read()
            max_limit = self._max_messages if max_messages is None else max_messages
            preserve = self._preserve_system if preserve_system is None else preserve_system
            if max_limit is not None:
                return sliding_window(msgs, max_limit, preserve)
            return msgs

    async def clear(self) -> None:
        async with self._lock:
            self._write([])

    def _read(self) -> list[ChatMessage]:
        raw = self._store.load()
        if not isinstance(raw, list):
            raise nRouterRequestError("MemoryStore.load() must return a list of messages.")
        return [_validate_message(m, f"MemoryStore.load()[{i}]") for i, m in enumerate(raw)]

    def _write(self, messages: list[ChatMessage]) -> None:
        self._store.save([_clone_message(m) for m in messages])


def create_array_store(seed: list[ChatMessage] | None = None) -> ArrayMemoryStore:
    return ArrayMemoryStore(seed)


def create_memory(
    store: MemoryStore | None = None,
    max_messages: int | None = None,
    preserve_system: bool = True,
) -> Memory:
    return Memory(store, max_messages=max_messages, preserve_system=preserve_system)


def _normalize_key(key: str) -> str:
    return key.lower().replace("_", "")


def _validate_message(message: Any, where: str) -> ChatMessage:
    if not isinstance(message, dict):
        raise nRouterRequestError(f"{where}: a message must be a dict.")
    for key in message:
        if _normalize_key(str(key)) in _TENANCY_KEYS:
            raise nRouterRequestError(
                f'{where}: a message must not carry the tenancy field "{key}".'
            )
    role = message.get("role")
    if role not in _ROLES:
        raise nRouterRequestError(
            f"{where}: role must be one of system, user, assistant, tool, developer."
        )
    content = message.get("content")
    tool_calls = message.get("tool_calls")
    has_tool_calls = isinstance(tool_calls, list) and len(tool_calls) > 0
    if content is None:
        if not has_tool_calls and role != "assistant":
            raise nRouterRequestError(f"{where}: content must be a string or content-parts list.")
    elif not isinstance(content, (str, list)):
        raise nRouterRequestError(f"{where}: content must be a string or content-parts list.")
    return _clone_message(message)


def _clone_message(message: ChatMessage) -> ChatMessage:
    return copy.deepcopy(message)

File: sdks/r/R/memory.R
#' Conversation Memory for nRouter
#'
#' Client-side conversation memory management with tenancy validation and sliding window pruning.
#'
#' @name nrouter-memory
NULL

TENANCY_KEYS <- c("organizationid", "orgid", "teamid", "userid", "nrouterorg")
VALID_ROLES <- c("system", "user", "assistant", "tool", "developer")

normalize_key <- function(key) {
  gsub("[-_]", "", tolower(key))
}

#' Validate a chat message
#'
#' @param message A named list representing a message.
#' @param context Context string for error reporting.
#' @return The validated message.
#' @export
nrouter_validate_message <- function(message, context = "nrouter_memory_add") {
  if (!is.list(message)) {
    stop(nrouter_configuration_condition(paste0(context, ": message must be a list")))
  }
  keys <- names(message)
  for (k in keys) {
    if (normalize_key(k) %in% TENANCY_KEYS) {
      stop(nrouter_configuration_condition(
        paste0(context, ": message contains forbidden tenancy key '", k, "'")
      ))
    }
  }
  role <- message$role
  if (is.null(role) || !is.character(role) || !(role %in% VALID_ROLES)) {
    stop(nrouter_configuration_condition(
      paste0(context, ": message must contain a valid role (system, user, assistant, tool, developer)")
    ))
  }
  content <- message$content
  tool_calls <- message$tool_calls
  has_tool_calls <- !is.null(tool_calls) && is.list(tool_calls) && length(tool_calls) > 0
  if (is.null(content)) {
    if (!has_tool_calls && role != "assistant") {
      stop(nrouter_configuration_condition(
        paste0(context, ": content must be a character string or list of content parts")
      ))
    }
  } else if (!is.character(content) && !is.list(content)) {
    stop(nrouter_configuration_condition(
      paste0(context, ": content must be a character string or list of content parts")
    ))
  }
  message
}

#' Sliding window pruning of message history
#'
#' @param messages List of messages.
#' @param max_messages Maximum messages to retain.
#' @param preserve_system Whether to preserve index 1 system/developer message.
#' @return Pruned list of messages.
#' @export
nrouter_sliding_window <- function(messages, max_messages = NULL, preserve_system = TRUE) {
  if (is.null(max_messages)) {
    return(messages)
  }
  if (!is.numeric(max_messages) || max_messages <= 0) {
    return(list())
  }
  limit <- as.integer(max_messages)
  n <- length(messages)
  if (n <= limit) {
    return(messages)
  }
  if (isTRUE(preserve_system) && n > 0) {
    first_role <- messages[[1]]$role
    if (!is.null(first_role) && (first_role == "system" || first_role == "developer")) {
      if (limit == 1L) {
        return(list(messages[[n]]))
      }
      tail_count <- limit - 1L
      start_idx <- n - tail_count + 1L
      return(c(list(messages[[1]]), messages[start_idx:n]))
    }
  }
  start_idx <- n - limit + 1L
  messages[start_idx:n]
}

#' In-memory array store for conversation messages
#'
#' @param seed Optional initial list of messages.
#' @return An environment with load and save functions.
#' @export
nrouter_create_array_store <- function(seed = list()) {
  store <- new.env(parent = emptyenv())
  store$messages <- seed

  load <- function() {
    store$messages
  }

  save <- function(messages) {
    store$messages <- messages
  }

  list(load = load, save = save)
}

#' Create conversation memory
#'
#' @param store Optional storage backend (defaults to in-memory array store).
#' @param max_messages Optional maximum messages limit for windowing.
#' @param preserve_system Whether to preserve system prompt during windowing.
#' @return An environment / list of memory methods (add, messages, clear).
#' @export
nrouter_create_memory <- function(store = NULL, max_messages = NULL, preserve_system = TRUE) {
  if (is.null(store)) {
    store <- nrouter_create_array_store()
  }

  add <- function(message) {
    clean <- nrouter_validate_message(message, context = "nrouter_memory$add")
    current <- store$load()
    current[[length(current) + 1L]] <- clean
    store$save(current)
    invisible(NULL)
  }

  messages <- function(max_msgs = max_messages, preserve_sys = preserve_system) {
    raw <- store$load()
    out <- lapply(seq_along(raw), function(i) {
      nrouter_validate_message(raw[[i]], context = paste0("MemoryStore$load()[[", i, "]]"))
    })
    if (!is.null(max_msgs) && max_msgs > 0) {
      nrouter_sliding_window(out, max_messages = max_msgs, preserve_system = preserve_sys)
    } else {
      out
    }
  }

  clear <- function() {
    store$save(list())
    invisible(NULL)
  }

  list(
    add = add,
    messages = messages,
    clear = clear
  )
}
